### ISSUE TITLE
关于VIT中位置编码的疑问self.pos_embed

### ISSUE THREAD

**Issue by d2623587501** [PaddlePaddle/PaddleClas#1772]
大佬您好，我在复现论文时发现，继承VisionTransformer类实例化位置编码self.pos_embed时后面加[:],如下代码，请问[]这是有什么含义吗

x += self.pos_embed[:]

**Comment by cuicheng01** [comment]
这个代码我没能没有看到，想问下是来自哪里呢？

**Comment by d2623587501** [comment]
感谢大佬的回复，以下是复现的源码
`class CustomVisionTransformer(VisionTransformer):
    def __init__(self, img_size=224, patch_size=16, *args, **kwargs):
        super(CustomVisionTransformer, self).__init__(img_size=img_size, patch_size=patch_size, *args, **kwargs)
        self.height, self.width = img_size
        self.patch_size = patch_size
    def forward_features(self, x):
        B, c, h, w = x.shape
        x = self.patch_embed(x)
        cls_tokens = self.cls_token.expand(B, -1, -1)  # stole cls_tokens impl from Phil Wang, thanks
        x = torch.cat((cls_tokens, x), dim=1)
        h, w = h//self.patch_size, w//self.patch_size
        pos_emb_ind = repeat(torch.arange(h)*(self.width//self.patch_size-w), 'h -> (h w)', w=w)+torch.arange(h*w)
        pos_emb_ind = torch.cat((torch.zeros(1), pos_emb_ind+1), dim=0).long()
        x += self.pos_embed[:, pos_emb_ind]
        #x = x + self.pos_embed
        x = self.pos_drop(x)
        for blk in self.blocks:
            x = blk(x)
        x = self.norm(x)
        return x`

通过继承VisionTransformer类，对图像进行编码，以下是复现代码

`class CustomVisionTransformer(VisionTransformer):
    def __init__(self, img_size=224, patch_size=16, *args, **kwargs):
        super(CustomVisionTransformer, self).__init__(img_size=img_size, patch_size=patch_size, *args, **kwargs)
        self.height, self.width = img_size
        self.patch_size = patch_size
    def forward_features(self, x):
        B, c, h, w = x.shape
        x = self.patch_embed(x)
        cls_tokens = self.cls_token
        cls_tokens = paddle.expand(cls_tokens,[B, -1, -1])  # stole cls_tokens impl from Phil Wang, thanks
        x = paddle.concat((cls_tokens, x), axis=1)
        print(x.shape)
        h, w = h//self.patch_size, w//self.patch_size
        pos_emb_ind_data = paddle.arange(h)*(self.width//self.patch_size-w)
        pos_emb_ind = paddle.sort(paddle.flatten(paddle.tile(pos_emb_ind_data, repeat_times=[w,1]),0), -1)+paddle.arange(h*w)
        pos_emb_ind = paddle.concat((paddle.zeros([1]), paddle.to_tensor(pos_emb_ind+paddle.to_tensor(1), dtype='float32')), axis=0)
        pos_emb_ind = paddle.to_tensor(pos_emb_ind, dtype='int32')
        #TODO
        x += self.pos_embed[:, pos_emb_ind]
        print(x)
        #x = x + self.pos_embed
        x = self.pos_drop(x)
        for blk in self.blocks:
            x = blk(x)
        x = self.norm(x)
        print(x)
        return x`

报错原因是 x += self.pos_embed[:, pos_emb_ind]其中x和self.pos_embed[:, pos_emb_ind]后的shape不一致，所有想请教以下，关于self.pos_embed[:, pos_emb_ind]的具体操作

**Comment by d2623587501** [comment]
上面代码有些紊乱，大佬您可以看一下项目中的代码
项目“transformer-ocr”共享链接(有效期三天)：https://aistudio.baidu.com/studio/project/partial/verify/3623588/a59d4fc5e11e4566b53fd32c9a78a7ab
paddle文件夹下的models_torch.py和models_paddle.py中的class CustomVisionTransformer(VisionTransformer)部分

**Comment by littletomatodonkey** [comment]
这个是对根据pos_emb_ind选定self.pos_embed中的特定列，你可以打印下选择前后，看下二者的shape区别

**Comment by d2623587501** [comment]
感谢大佬指点，问题已解决
